File: src/Tree/BinaryTreeBFSGuide/AverageofLevelsinBinaryTree/AverageofLevelsinBinaryTree.java
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
         List<Double> res = new LinkedList<>();
        
        if(root == null) return res;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        
        while(!q.isEmpty()){
            //Traverse level by level
            int size = q.size();
            double sum = 0;    
            for(int i = 0; i < size; i++){
                TreeNode first = q.poll();
                
                sum += (double) first.val;
                
                if(first.left != null){
                    q.add(first.left);
                }
                
                if(first.right != null){
                    q.add(first.right);
                }
            }
            double avg = sum / size; 
            res.add(avg);
        }
        return res;
    }
}

File: src/Tree/BinaryTreeBFSGuide/AverageofLevelsinBinaryTree/AverageofLevelsinBinaryTree.py
from collections import deque
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        res = []
        
        if root is None:
            return res
        
        queue = deque([root])
        
        while len(queue) != 0:
            size = len(queue)
            
            total = 0
            
            for idx in range(size):
                cur_node = queue.popleft()
                
                total += cur_node.val
                
                if cur_node.left is not None:
                    queue.append(cur_node.left)
                
                if cur_node.right is not None:
                    queue.append(cur_node.right)
                
            res.append(float(total / float(size * 1.0)))
            
        return res
